### PR TITLE
8252236: TabPane: must keep header of selected tab visible

### PR DESCRIPTION
**Pull request by kleopatra**
the issue is that the header of the selected tab is not always visible (or kept visible, see report for details)

- initially
- after changing side
- after resizing stage/tabPane
- after modifying the list of tabs

Fixed in TabPaneSkin to notify its TabHeaderArea (== collaborator that is responsible for layout the tabs) whenever its scroll might need an update.

Added tests that fail before and pass after the fix.

Note that there still are issues while/after dragging, which are separated into [JDK-8253352](https://bugs.openjdk.java.net/browse/JDK-8253352)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252236](https://bugs.openjdk.java.net/browse/JDK-8252236): TabPane: must keep header of selected tab visible

### Reviewers
 * [Ambarish Rapte](https://openjdk.java.net/census#arapte) (@arapte - **Reviewer**)
 * [Kevin Rushforth](https://openjdk.java.net/census#kcr) (@kevinrushforth - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/300/head:pull/300`
`$ git checkout pull/300`

### CODE REVIEW

**Comment by arapte** [comment]
In a bid to find more scenarios like you did mention description, I observed two more scenarios when selected tab does not stay in view.
1. Inserting tabs before or removing tabs after selected tab or moving the selected tab, causes the selected tab to go out of view.
2. Rotating TabPane can also take the selected tab out of view. I am little doubtful if it a legitimate issue. The rotations I tried clipped TabPane and TabPane header. 
I think rotation issue can be kept out of scope of this issue, can be filed as a follow on. I leave to it to you to take a call.

About 1, I did quickly verify that it gets fixed by adding `invalidateScrollOffset();` call in `TabHeaderArea.removeTab()`, `TabHeaderArea.addTab()` and  `TabHeaderArea.moveTab()`.
Also If we fix it here, then bug summary might need to change. It currently reflects only initial tab.

**Comment by kleopatra** [comment]
good catch and many thanks for the evaluation - now all I'll have to do is to write tests to catch them :) Will also update the bug summary as you suggest.

Agree, that the second can be regarded as off scope, as it seems to effect both header and content - will not do anything in this fix.

**Comment by kleopatra** [comment]
Widened scope of the issue to include problems after modifying the list of tabs: added fix and tests. 

Separated dragging issues into a new bug (there's an unrelated isssue when dragging over the leading edge, fixing that might have effects on the trailing edge, so didn't do much to fix the issue/s on the latter except ensuring the visibility after drop).

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java [hunk 1]
```diff
@@ -842,13 +847,13 @@ public TabHeaderArea() {
...
                         setScrollOffset(0.0);
                     } else {
-                        if (isSelectingTab) {
+                        if (scrollOffsetDirty) {
                             ensureSelectedTabIsVisible();
+                            scrollOffsetDirty = false;
                         } else {
-                            validateScrollOffset();
                         }
+                        validateScrollOffset();
```
Comment (arapte) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java:855]: This seems to be an unintended change. If so please revert.

**Review by arapte — COMMENTED** [review]
Looks good to me, left a minor comment.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java (hunk 1 above)
Comment (kleopatra) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java:855]: good eye :) Forgot to remove the empty else block.

Validate must be called always if the tabs are not fitting: needed if the last tab (in the list) is visible and we remove tabs from the end. Without, the now last tab is not kept glued to the trailing edge. There are two tests covering the corner case (testRemoveXXLast) which fail if we don't.

Question is if that should be mentioned in a code comment? Personally, I tend to just delete the empty block.

Comment (arapte) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java:855]: Thanks, I confirmed that it is a needed change. I think, a comment will be useful here, as the problem that is solved by this change is not very vivid from source.

Comment (kleopatra) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/TabPaneSkin.java]: added a comment to clarify what the validate is doing

**Review by arapte — APPROVED** [review]
Looks good to me.

**Review by kevinrushforth — APPROVED** [review]
The fix look good. Thanks for the additional, well-documented tests!
